tls/s2n_handshake_io.c

```diff
@@ -729,11 +797,16 @@ static int s2n_handshake_write_io(struct s2n_connection *conn)
      */
     if (s2n_stuffer_is_wiped(&conn->handshake.io)) {
         if (record_type == TLS_HANDSHAKE) {
-            GUARD(s2n_handshake_write_header(conn, ACTIVE_STATE(conn).message_type));
+            GUARD(s2n_handshake_write_header(&conn->handshake.io, ACTIVE_STATE(conn).message_type));
         }
         GUARD(ACTIVE_STATE(conn).handler[conn->mode] (conn));
         if (record_type == TLS_HANDSHAKE) {
-            GUARD(s2n_handshake_finish_header(conn));
+            GUARD(s2n_handshake_finish_header(&conn->handshake.io));
+        }
+
+        /* If we are sending out a HelloRetryRequest, we need to rebuild the hash transcript */
+        if (conn->handshake.message_number == SERVER_HELLO && conn->handshake.server_sent_hrr == 1) {
```

Comment:
could this actually be moved to post_handshake_update?

Reply:
`s2n_server_hello_retry_recreate_transcript` only recreates the synthetic message and the ClientHello1 hash. This let's the HelloRetryRequest message update the transcript during the usual call to `s2n_conn_update_handshake_hashes`. That's why I call this after the handler, but before the `s2n_conn_update_handshake_hashes`.